modules/sr/robot/utils.py

```diff
@@ -7,3 +20,75 @@ def map_to_range(
 ) -> float:
     """Maps a value from within one range of inputs to within a range of outputs."""
     return ((value - old_min) / (old_max - old_min)) * (new_max - new_min) + new_min
+
+
+def get_robot_emitter(robot: Robot, device: str) -> Emitter:
+    try:
+        # webots 2020b fails to assign the output to the expected type
+        emitter: Optional[Emitter] = robot.getDevice(device)
+        # webots 2020b always returns None when not raising TypeError
+        if emitter is None:
+            raise TypeError
+    except TypeError:
+        emitter = robot.getEmitter(device)
```

Comment:
This pattern looks very odd -- probably worth a comment about why we're calling `getEmitter` on `TypeError`.

Are there other sources of `TypeError` which this is catching, or is this just for our own one? If the latter, then I suggest simple branching might be neater. Otherwise a comment about the other source might be useful.

Reply:
getDevice seems to have rather odd behaviour in 2020b. On the one hand assigning the output to an untyped variable makes the variable Nonetype while assigning to type Optional[Emitter] throws an internal type error.
```
1|   File ".../competition-simulator/modules/sr/robot/motor_devices.py", line 14, in __init__
1|     self.webot_motor = get_robot_motor(webot, motor_name)
1|   File ".../competition-simulator/modules/sr/robot/utils.py", line 52, in get_robot_motor
1|     motor: Optional[Motor] = robot.getDevice(device)
1|   File "/usr/local/webots/lib/controller/python38/controller.py", line 2640, in getDevice
1|     if size == 0 or tag >= size:
1| TypeError: '>=' not supported between instances of 'str' and 'int'
```
Since I can't guarantee that the error will always occur I was just covvering my bases.
I'll add the type error information to the comment at the start of the try statement.

Reply:
I think it's worth thinking about the code flow as separate from the type annotations. In general type annotations don't have any effect at runtime, so we ought to see the same behaviour regardless of the annotations (if we're not then something very funky is going on).

It looks like in `2020b` the code in `getDevice` is expecting an `int`:
``` python
    def getDevice(tag):
      if tag == 0:
          return None
      size = len(Robot.__devices)
      if size == 0 or tag >= size:
          return None
      return Robot.__devices[tag]
```
What does the code look like in 2021a?

----
I think it might help clarity if we condense the `try...except TypeError` around just the part of the code which we know will raise `TypeError` and move our own `None` check outside that. If we're primarily catching `TypeError` to work around the version differences then it may make sense to have our own helper which does literally just the version differences handling (but none of the type annotation handling) which we can then wrap in handlers which do the type specific stuff.

Reply:
2021a seems to run similar code
```python
%rename ("__internalGetDeviceTagFromName") webots::Robot::getDeviceTagFromName;
%extend webots::Robot {
  %pythoncode %{
def getDevice(self, name):
  tag = self.__internalGetDeviceTagFromName(name)
  return self.__getOrCreateDevice(tag)
@staticmethod
def internalGetDeviceFromTag(tag):
  if tag == 0:
      return None
  size = len(Robot.__devices)
  if size == 0 or tag >= size:
      return None
  return Robot.__devices[tag]
```
I've changed the error handling code to the recommendation in the comment below

Reply:
Note the call to `__internalGetDeviceTagFromName` to convert the incoming parameter to `getDevice` from a `name` into a `tag` though -- I bet that's the fix.

Your new code looks good though :+1:. (I'll have a better look this evening)